pinot-controller/src/main/java/com/linkedin/pinot/controller/util/TableRetentionValidator.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.util;
+
+import com.linkedin.pinot.common.config.AbstractTableConfig;
+import com.linkedin.pinot.common.config.SegmentsValidationAndRetentionConfig;
+import com.linkedin.pinot.common.metadata.segment.OfflineSegmentZKMetadata;
+import com.linkedin.pinot.common.utils.time.TimeUtils;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import org.apache.helix.PropertyPathConfig;
+import org.apache.helix.PropertyType;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZKHelixAdmin;
+import org.apache.helix.manager.zk.ZNRecordSerializer;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The <code>TableRetentionValidator</code> class validates the retention policy in table config, and the start/end
+ * timestamp in segment metadata.
+ * <p>Will validate the followings:
+ * <ul>
+ *   <li>
+ *     Table Config
+ *     <ul>
+ *       <li>"segmentsConfig" is set.</li>
+ *       <li>"segmentPushType" is set to APPEND or REFRESH.</li>
+ *       <li>Retention setting is valid for APPEND push type.</li>
+ *     </ul>
+ *   </li>
+ *   <li>
+ *     Segment Metadata
+ *     <ul>
+ *       <li>For APPEND push type, offline segment start/end time and time unit is valid.</li>
+ *     </ul>
+ *   </li>
+ * </ul>
+ */
+public class TableRetentionValidator {
```

Comment:
The reason I put it in controller is that controller can integrate this into the validator in the future. When you run it from PinotToolLauncher, it can log properly.